src/Compilers/Core/MSBuildTask/DotnetHost.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    internal sealed class DotnetHost
+    {
+        public string ToolNameOpt { get; }
+
+        public string PathToToolOpt { get; }
+
+        public string CommandLineArgs { get; }
+
+        private DotnetHost(string toolNameOpt, string pathToToolOpt, string commandLineArgs)
+        {
+            ToolNameOpt = toolNameOpt;
+            PathToToolOpt = pathToToolOpt;
+            CommandLineArgs = commandLineArgs;
+        }
+
+        public static DotnetHost CreateUnmanagedToolInvocation(string pathToTool, string commandLineArgs)
+        {
+            return new DotnetHost(null, pathToTool, commandLineArgs);
+        }
+
+        public static DotnetHost CreateManagedToolInvocation(string toolNameWithoutExtension, string commandLineArgs)
+        {
+            string pathToToolOpt;
+            string toolName;
+            if (IsCliHost(out string pathToDotnet))
+            {
+                toolName = $"{toolNameWithoutExtension}.dll";
+                pathToToolOpt = Utilities.GenerateFullPathToTool(toolName);
+                if (pathToToolOpt != null)
+                {
+                    commandLineArgs = $"\"{pathToToolOpt}\" {commandLineArgs}";
```

Comment:
> `pathToToolOpt` is already quoted

This seems a little odd from a nitpicky point of view - there's no way to tell the difference between a file named literally `"file"` and a quoted string of the name `file`.

I agree with better quoting, though - escaping backslashes, as well as internal quotes in the filename.